feat(Topology/Compactness): first-countable locally path-connected spaces are delta-generated

peabrainiac opened the pull request:
Shows that all first-countable locally path-connected spaces are delta-generated (so in particular all normed spaces and convex subsets thereof are), and that delta-generated spaces are equivalently generated by the unit interval or standard simplices.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->
The main lemma used in this construction is unfortunately quite messy - it requires concatenating countably many paths leading up to a point, which is easy to see on paper but hard to formalise in lean. I've documented the argument in `LocPathConnectedSpace.eq_coinduced_paths` as good as I could, but eventually gave up on getting the `Path.sigmaConcat` construction into a good state and just marked it as private.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

YaelDillies commented on Mathlib/Topology/Compactness/DeltaGeneratedSpace.lean:
```diff
@@ -181,3 +180,247 @@ instance Sum.deltaGeneratedSpace [DeltaGeneratedSpace X] [DeltaGeneratedSpace Y]
+    (γ : (n : ℕ) → Path (s n) (s n.succ)) (x : X) : I → X := fun t ↦ by
+  refine if ht : t < 1 then ?_ else x
+  have ht' : 0 < σ t := by rw [lt_symm_comm]; simp [ht]
+  let n := Nat.floor (- Real.logb 2 (σ t))
```
Can you use the notation?
```suggestion
  let n := ⌊- Real.logb 2 (σ t)⌋₊
```

YaelDillies commented on Mathlib/Topology/Compactness/DeltaGeneratedSpace.lean:
```diff
@@ -4,6 +4,8 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Ben Eltschig
 -/
 import Mathlib.Analysis.LocallyConvex.WithSeminorms
+import Mathlib.Topology.ContinuousMap.Interval
+import Mathlib.Analysis.SpecialFunctions.Log.Base
```
Can you keep imports sorted?
```suggestion
import Mathlib.Analysis.LocallyConvex.WithSeminorms
import Mathlib.Analysis.SpecialFunctions.Log.Base
import Mathlib.Topology.ContinuousMap.Interval
```

YaelDillies commented on Mathlib/Topology/Compactness/DeltaGeneratedSpace.lean:
```diff
@@ -181,3 +180,247 @@ instance Sum.deltaGeneratedSpace [DeltaGeneratedSpace X] [DeltaGeneratedSpace Y]
+  corresponding path is defined separately because continuity is annoying to prove. -/
+private noncomputable def Path.sigmaConcatFun {X : Type*} [TopologicalSpace X] {s : ℕ → X}
+    (γ : (n : ℕ) → Path (s n) (s n.succ)) (x : X) : I → X := fun t ↦ by
+  refine if ht : t < 1 then ?_ else x
+  have ht' : 0 < σ t := by rw [lt_symm_comm]; simp [ht]
+  let n := Nat.floor (- Real.logb 2 (σ t))
+  let t' : I := ⟨2 * (1 - σ t * 2 ^ (n : ℝ)), by
+    have hn : (n : ℝ) ≤ _ := Nat.floor_le (a := - Real.logb 2 (σ t)) <|
+      Left.nonneg_neg_iff.2 <| Real.logb_nonpos one_lt_two (σ t).2.1 (σ t).2.2
+    have h : (2 : ℝ) ^ (n : ℝ) ≤ (σ t).1 ⁻¹ := by
+      refine (Real.rpow_le_rpow_of_exponent_le one_le_two hn).trans ?_
+      rw [Real.rpow_neg zero_le_two, Real.rpow_logb two_pos (by simp) (by simp [ht])]
+    have h' : (σ t) * (2 : ℝ) ^ (n : ℝ) ≤ 1 := by
+      refine (mul_le_mul_of_nonneg_left h (σ t).2.1).trans (mul_inv_cancel₀ ?_).le
+      exact (coe_pos.2 ht').ne.symm
+    linarith, by
+    have hn : _ < (n.succ : ℝ) := Nat.lt_succ_floor _
+    have h : (σ t).1⁻¹ ≤ 2 * 2 ^ (n : ℝ) := by
+      rw [Real.rpow_natCast, ← pow_succ', ← Real.rpow_natCast]
+      refine (Eq.le ?_).trans <| Real.rpow_le_rpow_of_exponent_le one_le_two hn.le
+      rw [Real.rpow_neg zero_le_two, Real.rpow_logb two_pos (by simp) (by simp [ht])]
+    suffices h : 1 ≤ (σ t) * (2 * (2 : ℝ) ^ (n : ℝ)) by rw [mul_left_comm] at h; linarith
+    refine (mul_inv_cancel₀ ?_).symm.le.trans <| mul_le_mul_of_nonneg_left h (σ t).2.1
+    exact (coe_pos.2 ht').ne.symm⟩
+  exact γ n t'
```
Here's a better way to present this definition
```suggestion
  let n := ⌊- Real.logb 2 (σ t)⌋₊
  refine if ht : t < 1 then γ n ⟨2 * (1 - σ t * 2 ^ (n : ℝ)), ?_, ?_⟩ else x
  · have ht' : 0 < σ t := by rw [lt_symm_comm]; simp [ht]
    have hn : (n : ℝ) ≤ _ := Nat.floor_le (a := - Real.logb 2 (σ t)) <|
      Left.nonneg_neg_iff.2 <| Real.logb_nonpos one_lt_two (σ t).2.1 (σ t).2.2
    have h : (2 : ℝ) ^ (n : ℝ) ≤ (σ t).1 ⁻¹ := by
      refine (Real.rpow_le_rpow_of_exponent_le one_le_two hn).trans ?_
      rw [Real.rpow_neg zero_le_two, Real.rpow_logb two_pos (by simp) (by simp [ht])]
    have h' : (σ t) * (2 : ℝ) ^ (n : ℝ) ≤ 1 := by
      refine (mul_le_mul_of_nonneg_left h (σ t).2.1).trans (mul_inv_cancel₀ ?_).le
      exact (coe_pos.2 ht').ne.symm
    linarith
  · have hn : _ < (n.succ : ℝ) := Nat.lt_succ_floor _
    have ht' : 0 < σ t := by rw [lt_symm_comm]; simp [ht]
    have h : (σ t).1⁻¹ ≤ 2 * 2 ^ (n : ℝ) := by
      rw [Real.rpow_natCast, ← pow_succ', ← Real.rpow_natCast]
      refine (Eq.le ?_).trans <| Real.rpow_le_rpow_of_exponent_le one_le_two hn.le
      rw [Real.rpow_neg zero_le_two, Real.rpow_logb two_pos (by simp) (by simp [ht])]
    suffices h : 1 ≤ (σ t) * (2 * (2 : ℝ) ^ (n : ℝ)) by rw [mul_left_comm] at h; linarith
    refine (mul_inv_cancel₀ ?_).symm.le.trans <| mul_le_mul_of_nonneg_left h (σ t).2.1
    exact (coe_pos.2 ht').ne.symm
```

YaelDillies commented on Mathlib/Topology/Compactness/DeltaGeneratedSpace.lean:
```diff
@@ -181,3 +180,247 @@ instance Sum.deltaGeneratedSpace [DeltaGeneratedSpace X] [DeltaGeneratedSpace Y]
+private lemma Path.sigmaConcatFun_eqOn {X : Type*} [TopologicalSpace X] {s : ℕ → X}
+    (γ : (n : ℕ) → Path (s n) (s n.succ)) {x : X} (n : ℕ) :
+    Set.EqOn (sigmaConcatFun γ x) (fun t ↦ (γ n).extend (2 * (1 - (1 - t) * (2 ^ n))))
+    (Set.Icc (σ ⟨1 / 2 ^ n, by simp [inv_le_one₀, one_le_pow₀]⟩)
+      (σ ⟨1 / 2 ^ (n+1), by simp [inv_le_one₀, one_le_pow₀]⟩)) := fun t ht ↦ by
```
Can you avoid using `1 / _`?
```suggestion
    (Set.Icc (σ ⟨2⁻¹  ^ n, by simp [inv_le_one₀, one_le_pow₀]⟩)
      (σ ⟨2⁻¹ ^ (n+1), by simp [inv_le_one₀, one_le_pow₀]⟩)) := fun t ht ↦ by
  simp only [Set.mem_Icc, ← Subtype.coe_le_coe, coe_symm_eq] at ht
```

YaelDillies commented on Mathlib/Topology/Compactness/DeltaGeneratedSpace.lean:
```diff
@@ -181,3 +180,247 @@ instance Sum.deltaGeneratedSpace [DeltaGeneratedSpace X] [DeltaGeneratedSpace Y]
+    simp [hn, pow_succ]
+    linear_combination ht''
+
+/-- The concatenation of countably many paths leading up to some point `x`. -/
+private noncomputable def Path.sigmaConcat {X : Type*} [TopologicalSpace X] {s : ℕ → X}
+    (γ : (n : ℕ) → Path (s n) (s n.succ)) (x : X) {b : ℕ → Set X} (hb : (𝓝 x).HasAntitoneBasis b)
+    (hγ : ∀ n t, γ n t ∈ b n) : Path (s 0) x where
+  toFun := sigmaConcatFun γ x
```
This looks like a great definition. Why keep it private? I think it should live in an earlier file too (possibly a new file)

YaelDillies commented on Mathlib/Topology/Compactness/DeltaGeneratedSpace.lean:
```diff
@@ -181,3 +180,247 @@ instance Sum.deltaGeneratedSpace [DeltaGeneratedSpace X] [DeltaGeneratedSpace Y]
+
+/-- The concatenation of countably many paths leading up to some point `x`. -/
+private noncomputable def Path.sigmaConcat {X : Type*} [TopologicalSpace X] {s : ℕ → X}
+    (γ : (n : ℕ) → Path (s n) (s n.succ)) (x : X) {b : ℕ → Set X} (hb : (𝓝 x).HasAntitoneBasis b)
```
It feels a bit weird to ask for an antitone basis when in fact an *eventually* antitone basis is enough. I want to say something like "the paths uniformly converge towards the point", but I'm not sure how to say that without the data of a uniformity

YaelDillies commented on Mathlib/Topology/Compactness/DeltaGeneratedSpace.lean:
```diff
@@ -181,3 +180,247 @@ instance Sum.deltaGeneratedSpace [DeltaGeneratedSpace X] [DeltaGeneratedSpace Y]
+          le_sub_comm.1 ht.2).trans <| by simp [- Nat.cast_add, Real.logb_pow]
+      · exact (Eq.le (by simp [Real.logb_pow])).trans <| Nat.floor_le_floor <|
+          neg_le_neg <| Real.logb_le_logb_of_le one_lt_two ht'' <| sub_le_comm.1 ht.1
+    have ht'' : (2 * (1 - (1 - t.1) * 2 ^ n)) = 1 := by
```
```suggestion
    have ht'' : 2 * (1 - (1 - t.1) * 2 ^ n) = 1 := by
```

kbuzzard commented:
We've just had a look at this PR at the triage party; here are some comments.
1) This work looks great; I wonder if some of the private lemmas should in fact be public (specifically the one gluing infinitely many paths to form a path)
2) The large imports are nothing to worry about; this is basically a leaf file and if the logarithm is useful here, that's great.
3) Some of the proofs might need some rewriting but only in a minor way; it was a bit jarring to see a definition with a very long proof in the middle when the goal was still a type, but probably this doesn't actually matter.
4) I am always scared of `convert; [several more lines of proof]` because such proofs can be hard to maintain later on.

peabrainiac commented:
In addition to incorporating some of the smaller suggestions, I have replaced my previous usage of `⌊Real.logb 2 x⌋₊` with `Nat.log 2 ⌊x⌋₊` following a suggestion by Yael on discord. This has cut down on the large imports significantly - while I still rely on analysis files for `tendsto_pow_atTop_nhds_zero_of_abs_lt_one` in `LocPathConnectedSpace.eq_coinduced_paths`, the lemmas regarding `Path.sigmaConcat` now depend only on `Data.Nat.Log`, and should thus be fine to move to e.g. `Mathlib.Topology.Connected.PathConnected`. I will try that in a few hours when I have time.

YaelDillies commented:
(for the record, since the bot threw away its original message, the previous import increase was +188. It now is +61)

peabrainiac commented:
I have changed `Path.sigmaConcat` and the lemmas surrounding it from private to public, moved them to `PathConnected.lean` and also added a lemma `Path.sigmaConcat_eq_trans` relating `Path.sigmaConcat` to simple concatenation of paths, since I felt that was too important to be left out if this is meant to be public API.

However, doing that has pushed `PathConnected.lean` beyond the 1500 line limit, so I have opened another PR #21591 to split it up; it should be ready for review now. I will come back to this PR tomorrow to hopefully also get it into a final state.

YaelDillies commented on Mathlib/Topology/ContinuousOn.lean:
```diff
@@ -1495,6 +1495,17 @@ lemma ContinuousOn.union_continuousAt
+theorem ContinuousOn.union_isClosed {X Y : Type*} [TopologicalSpace X] [TopologicalSpace Y]
+    {s t : Set X} (hs : IsClosed s) (ht : IsClosed t) {f : X → Y} (hfs : ContinuousOn f s)
+    (hft : ContinuousOn f t) : ContinuousOn f (s ∪ t) := by
+  refine fun x hx ↦ ContinuousWithinAt.union ?_ ?_
```
```suggestion
  refine fun x hx ↦ .union ?_ ?_
```

YaelDillies commented on Mathlib/Topology/ContinuousOn.lean:
```diff
@@ -1495,6 +1495,17 @@ lemma ContinuousOn.union_continuousAt
+    (hft : ContinuousOn f t) : ContinuousOn f (s ∪ t) := by
+  refine fun x hx ↦ ContinuousWithinAt.union ?_ ?_
+  · refine if hx : x ∈ s then hfs x hx else continuousWithinAt_of_not_mem_closure ?_
+    rw [hs.closure_eq]; exact hx
+  · refine if hx : x ∈ t then hft x hx else continuousWithinAt_of_not_mem_closure ?_
+    rw [ht.closure_eq]; exact hx
```
```suggestion
    rwa [hs.closure_eq]
  · refine if hx : x ∈ t then hft x hx else continuousWithinAt_of_not_mem_closure ?_
    rwa [ht.closure_eq]
```

YaelDillies commented on Mathlib/Topology/ContinuousOn.lean:
```diff
@@ -1495,6 +1495,17 @@ lemma ContinuousOn.union_continuousAt
+/-- If a function is continuous on two closed sets, it is also continuous on their union. -/
+theorem ContinuousOn.union_isClosed {X Y : Type*} [TopologicalSpace X] [TopologicalSpace Y]
+    {s t : Set X} (hs : IsClosed s) (ht : IsClosed t) {f : X → Y} (hfs : ContinuousOn f s)
+    (hft : ContinuousOn f t) : ContinuousOn f (s ∪ t) := by
```
I'm having a hard time thinking we don't have this, but I can't find it even with loogle. Can you make it a separate PR so as to lure people like Patrick into reviewing?

YaelDillies commented on Mathlib/Topology/Connected/PathConnected.lean:
```diff
@@ -1206,6 +1388,17 @@ theorem isOpen_isPathConnected_basis (x : X) :
     isPathConnected_pathComponentIn hxu⟩⟩, pathComponentIn_subset.trans hus⟩
 
+/-- In a first-countable locally path-connected space, every point has a neighbourhood basis
+  that is a decreasing sequence of path-connected sets. -/
```
Please don't indent docstrings by default
```suggestion
that is a decreasing sequence of path-connected sets. -/
```

YaelDillies commented on Mathlib/Topology/Connected/PathConnected.lean:
```diff
@@ -656,6 +657,187 @@ theorem refl_reparam {f : I → I} (hfcont : Continuous f) (hf₀ : f 0 = 0) (hf
+        rw [← Nat.pow_le_iff_le_log one_lt_two (Nat.floor_pos.2 <| (one_le_inv₀ ht''').2
+          (σ t).2.2).ne.symm, Nat.le_floor_iff (inv_pos.2 ht''').le, le_inv_comm₀ (by simp) ht''',
+          coe_symm_eq, sub_le_comm]
+        apply LT.lt.le; simpa using ht
```
```suggestion
        simpa using ht.le
```

peabrainiac replied:
This doesn't work, and `simpa using (Set.mem_Ioi.1 ht).le`  doesn't either; I'm not really sure why, but I'm guessing whatever would work isn't going to be shorter than `apply LT.lt.le; simpa using ht`, so I'm keeping this line as is for now.

YaelDillies commented on Mathlib/Topology/Connected/PathConnected.lean:
```diff
@@ -656,6 +657,187 @@ theorem refl_reparam {f : I → I} (hfcont : Continuous f) (hf₀ : f 0 = 0) (hf
 
+/-! #### Concatenating countably many paths -/
+
+
```
```suggestion
```

YaelDillies commented on Mathlib/Topology/Connected/PathConnected.lean:
Can this too become a separate PR?
